crates/hir/src/semantics/source_to_def.rs

```diff
@@ -1,4 +1,73 @@
 //! Maps *syntax* of various definitions to their semantic ids.
+//!
+//! This is a very interesting module, and, in some sense, can be considered a
+//! heart of the IDE parts of rust-analyzer.
+//!
+//! This module solves the following problem:
+//!
+//!     Given a piece of syntax, find the corresponding semantic definition (def).
+//!
+//! This problem is a part of more-or-less every IDE feature implemented. Every
+//! IDE functionality (like goto to definition), conceptually starts with a
+//! specific cursor position in a file. Starting with this text offset, we first
+//! figure out what syntactic construct are we at: is this a pattern, an
+//! expression, an item definition.
+//!
+//! Knowing only the syntax gives us relatively little info. For example,
+//! looking at the syntax of the function we can realise that it is a part of an
+//! `impl` block, but we won't be able to tell what trait function the current
+//! function overrides, and whether it does that correctly. For that, we need to
+//! go from [`ast::Fn`] to [`crate::Function], and that's exactly what this
+//! module does.
+//!
+//! As syntax trees are values and don't know their place of origin/identity,
+//! this module also requires [`InFile`] wrappers to understand which specific
+//! real or macro-expanded file the tree comes from.
```

Comment:
it might be interesting to note that this choice of file is already a 'semantic' decision (if there are multiple crates using the same file, this already determines the crate) and provides the base case for the recursion

Reply:
> Edit: Ah wait, it doesn't, doesn't it? only if it's a macro file thinking

Makes me wonder if that "half-semanticness" of `HirFileId` is something we should change? It's kind of weird that a `HirFileId` belongs to a specific crate, but only if it's a macro call, isn't it?

Reply:
Good point, looks like I've just forgot to mention the inherent ambiguity we have here....

Reply:
Lol, half-semantics is my new favorite euphemism :rofl: 

Yeah, I think there are two problems with `HirFileId` / `InFile`:

* it's a poor abstraction to expose to the IDE. That probably is part of #8713. I think new hir nodes will get `.location()` method, which will return an *opaque* location, which designates a real span, a macro span, an ether span (for things like i32) or a "got the stuff from .rlib" span.
* it's just buggy internally. `fn parse(&self, file_id: FileId) -> Parse<ast::SourceFile>` can't take edition into account. 

It generally feels as if we should start moving towards tree :fearful: